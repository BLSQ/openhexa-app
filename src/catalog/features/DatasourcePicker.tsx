import { gql, useQuery } from "@apollo/client";
import Select, {
  SelectOption,
  SelectProps,
} from "core/components/forms/Select";
import { useTranslation } from "next-i18next";
import { useMemo } from "react";
import { DatasourcePickerQuery } from "./DatasourcePicker.generated";

type Option = SelectOption & { symbol?: string; app?: string };

type DatasourcePickerProps = {
  onChange(values: Option[] | null): void;
  value: any;
  className?: string;
} & Omit<
  SelectProps<Option>,
  "onChange" | "options" | "value" | "getOptionLabel"
>;

const DatasourcePicker = (props: DatasourcePickerProps) => {
  const { t } = useTranslation();
  const {
    value,
    onChange,
    placeholder = t("Select datasources"),
    ...delegated
  } = props;

  const { data, loading } = useQuery<DatasourcePickerQuery>(gql`
    query DatasourcePicker {
      catalog(page: 1, perPage: 25) {
        items {
          id
          objectId
          name
          symbol
          type {
            name
            app
          }
        }
      }
    }
  `);

  const options = useMemo(
    () =>
      data?.catalog.items.map((item) => ({
        id: item.objectId,
        symbol: item.symbol,
        label: item.name,
        app: item.type.app,
        type: item.type.name,
      })) ?? [],
    [data]
  );

  const getOptionLabel = (option: Option) => {
    return (
      <div className="flex gap-2">
        {option.symbol && (
          <div className="h-5 w-5 flex-shrink-0">
            <img src={option.symbol} alt="" />
          </div>
        )}
        <span>{option.label}</span>
      </div>
    );
  };
  return (
    <Select<Option>
      {...delegated}
      placeholder={placeholder ?? undefined}
      value={value}
      onChange={onChange}
      options={options}
      getOptionLabel={getOptionLabel}
    />
  );
};

export default DatasourcePicker;
