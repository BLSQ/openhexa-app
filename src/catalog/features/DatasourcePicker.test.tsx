import DatasourcePicker from "./DatasourcePicker";
import { render } from "@testing-library/react";

import { MockedProvider } from "@apollo/client/testing";

describe("DatasourcePicker", () => {
  it("renders", async () => {
    const onChange = jest.fn();

    const { container } = render(
      <MockedProvider mocks={[]}>
        <DatasourcePicker onChange={onChange} value={undefined} />
      </MockedProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
