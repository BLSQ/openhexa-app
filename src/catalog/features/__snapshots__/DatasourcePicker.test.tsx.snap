// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasourcePicker renders 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      class="form-input flex w-full items-center rounded-md border-gray-300 shadow-sm disabled:border-gray-300 focus-within:outline-none focus:ring-transparent focus-visible:border-blue-500 disabled:cursor-not-allowed  sm:text-sm hover:border-gray-400"
    >
      <div
        class="mr-1 flex flex-1 items-center truncate"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          autocomplete="off"
          class="flex-1 placeholder-gray-600 placeholder-opacity-70 outline-none"
          data-headlessui-state=""
          data-testid="combobox-input"
          id="headlessui-combobox-input-6fbe024f-2316-4265-a6e8-d65a837e308a"
          placeholder="Select datasources"
          role="combobox"
          type="text"
        />
      </div>
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        data-headlessui-state=""
        data-testid="combobox-button"
        id="headlessui-combobox-button-e67862f3-ccbf-4c51-b8ed-ed1d0420ea19"
        tabindex="-1"
        type="button"
      >
        <div
          class="ml-1 flex items-center gap-0.5 rounded-r-md text-gray-400 focus:outline-none"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5 hover:text-gray-500"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 15L12 18.75 15.75 15m-7.5-6L12 5.25 15.75 9"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</div>
`;
